File: ports/typescript/src/sha3-kmac.ts
import { cshake128, cshake256 } from '@noble/hashes/sha3';

/**
 * Left encode function for KMAC
 * Encodes an integer as a byte string with length prefix on the left
 */
function leftEncode(x: number): Uint8Array {
  if (x === 0) {
    return new Uint8Array([1, 0]);
  }

  const temp = new Uint8Array(8);
  let n = 0;

  for (let i = 7; i >= 0; i--) {
    temp[i] = x & 0xff;
    x = Math.floor(x / 256);
  }

  let startIdx = 0;
  while (startIdx < 8 && temp[startIdx] === 0) {
    startIdx++;
  }
  n = 8 - startIdx;

  const result = new Uint8Array(n + 1);
  result[0] = n;
  result.set(temp.slice(startIdx), 1);
  return result;
}

/**
 * Right encode function for KMAC
 * Encodes an integer as a byte string with length suffix on the right
 */
function rightEncode(x: number): Uint8Array {
  if (x === 0) {
    return new Uint8Array([0, 1]);
  }

  const temp = new Uint8Array(8);
  let n = 0;

  for (let i = 7; i >= 0; i--) {
    temp[i] = x & 0xff;
    x = Math.floor(x / 256);
  }

  let startIdx = 0;
  while (startIdx < 8 && temp[startIdx] === 0) {
    startIdx++;
  }
  n = 8 - startIdx;

  const result = new Uint8Array(n + 1);
  result.set(temp.slice(startIdx), 0);
  result[n] = n;
  return result;
}

/**
 * Encode a byte string with its bit length
 */
function encodeString(data: Uint8Array): Uint8Array {
  const bitLen = data.length * 8;
  const encoded = leftEncode(bitLen);

  const result = new Uint8Array(encoded.length + data.length);
  result.set(encoded, 0);
  result.set(data, encoded.length);
  return result;
}

/**
 * Bytepad function for KMAC
 * Prepends an encoding of the rate w and pads with zeros
 */
function bytepad(data: Uint8Array, w: number): Uint8Array {
  const wEncoded = leftEncode(w);
  const totalLen = wEncoded.length + data.length;

  let padLen = w - (totalLen % w);
  if (padLen === w) {
    padLen = 0;
  }

  const result = new Uint8Array(totalLen + padLen);
  result.set(wEncoded, 0);
  result.set(data, wEncoded.length);
  return result;
}

/**
 * KMAC implementation using cSHAKE
 */
function kmac(
  key: Uint8Array,
  customization: Uint8Array,
  data: Uint8Array,
  outputLen: number,
  rate: number,
  cshakeFunc: (data: Uint8Array, opts: { personalization: Uint8Array; NISTfn: Uint8Array; dkLen: number }) => Uint8Array
): Uint8Array {
  const encodedKey = encodeString(key);
  const paddedKey = bytepad(encodedKey, rate);

  // Concatenate paddedKey, data, and right_encode(outputLen * 8)
  const rightEncoded = rightEncode(outputLen * 8);
  const inputData = new Uint8Array(paddedKey.length + data.length + rightEncoded.length);
  inputData.set(paddedKey, 0);
  inputData.set(data, paddedKey.length);
  inputData.set(rightEncoded, paddedKey.length + data.length);

  return cshakeFunc(inputData, {
    personalization: customization,
    NISTfn: new TextEncoder().encode('KMAC'),
    dkLen: outputLen,
  });
}

/**
 * SHA3-KMAC128 for passcode (internal use)
 */
export function sha3KMAC128ForPasscode(key: Uint8Array, data: Uint8Array): Uint8Array {
  return kmac(key, new TextEncoder().encode('authorization'), data, 32, 168, cshake128);
}

/**
 * SHA3-KMAC128 with customizable parameters
 */
export function sha3KMAC128(key: Uint8Array, customization: Uint8Array, data: Uint8Array, outputLen: number): Uint8Array {
  return kmac(key, customization, data, outputLen, 168, cshake128);
}

/**
 * SHA3-KMAC256 for passcode (internal use)
 */
export function sha3KMAC256ForPasscode(key: Uint8Array, data: Uint8Array): Uint8Array {
  return kmac(key, new TextEncoder().encode('authorization'), data, 32, 136, cshake256);
}

/**
 * SHA3-KMAC256 with customizable parameters
 */
export function sha3KMAC256(key: Uint8Array, customization: Uint8Array, data: Uint8Array, outputLen: number): Uint8Array {
  return kmac(key, customization, data, outputLen, 136, cshake256);
}

File: ports/rust/src/sha3_kmac.rs
use sha3::digest::{ExtendableOutput, Update, XofReader};
use sha3::{CShake128, CShake256};

/// Left encode function for KMAC
fn left_encode(x: u64) -> Vec<u8> {
    if x == 0 {
        return vec![1, 0];
    }

    let mut temp = [0u8; 8];
    let mut val = x;
    
    for i in (0..8).rev() {
        temp[i] = (val & 0xff) as u8;
        val >>= 8;
    }

    let mut start_idx = 0;
    while start_idx < 8 && temp[start_idx] == 0 {
        start_idx += 1;
    }
    let n = 8 - start_idx;

    let mut result = Vec::with_capacity(n + 1);
    result.push(n as u8);
    result.extend_from_slice(&temp[start_idx..]);
    result
}

/// Right encode function for KMAC
fn right_encode(x: u64) -> Vec<u8> {
    if x == 0 {
        return vec![0, 1];
    }

    let mut temp = [0u8; 8];
    let mut val = x;
    
    for i in (0..8).rev() {
        temp[i] = (val & 0xff) as u8;
        val >>= 8;
    }

    let mut start_idx = 0;
    while start_idx < 8 && temp[start_idx] == 0 {
        start_idx += 1;
    }
    let n = 8 - start_idx;

    let mut result = Vec::with_capacity(n + 1);
    result.extend_from_slice(&temp[start_idx..8]);
    result.push(n as u8);
    result
}

/// Encode a byte string with its bit length
fn encode_string(data: &[u8]) -> Vec<u8> {
    let bit_len = (data.len() * 8) as u64;
    let encoded = left_encode(bit_len);

    let mut result = Vec::with_capacity(encoded.len() + data.len());
    result.extend_from_slice(&encoded);
    result.extend_from_slice(data);
    result
}

/// Bytepad function for KMAC
fn bytepad(data: &[u8], w: usize) -> Vec<u8> {
    let w_encoded = left_encode(w as u64);
    let total_len = w_encoded.len() + data.len();

    let mut pad_len = w - (total_len % w);
    if pad_len == w {
        pad_len = 0;
    }

    let mut result = Vec::with_capacity(total_len + pad_len);
    result.extend_from_slice(&w_encoded);
    result.extend_from_slice(data);
    result.resize(total_len + pad_len, 0);
    result
}

/// KMAC implementation using CShake128
fn kmac128(
    key: &[u8],
    customization: &[u8],
    data: &[u8],
    output_len: usize,
) -> Vec<u8> {
    let encoded_key = encode_string(key);
    let padded_key = bytepad(&encoded_key, 168); // rate for SHA3-128

    // NIST SP 800-185: KMAC uses cSHAKE with function name "KMAC" and customization
    let mut hasher = CShake128::from_core(
        sha3::CShake128Core::new_with_function_name(b"KMAC", customization),
    );
    
    hasher.update(&padded_key);
    hasher.update(data);
    hasher.update(&right_encode((output_len * 8) as u64));

    let mut output = vec![0u8; output_len];
    hasher.finalize_xof().read(&mut output);
    output
}

/// KMAC implementation using CShake256
fn kmac256(
    key: &[u8],
    customization: &[u8],
    data: &[u8],
    output_len: usize,
) -> Vec<u8> {
    let encoded_key = encode_string(key);
    let padded_key = bytepad(&encoded_key, 136); // rate for SHA3-256

    // NIST SP 800-185: KMAC uses cSHAKE with function name "KMAC" and customization
    let mut hasher = CShake256::from_core(
        sha3::CShake256Core::new_with_function_name(b"KMAC", customization),
    );
    
    hasher.update(&padded_key);
    hasher.update(data);
    hasher.update(&right_encode((output_len * 8) as u64));

    let mut output = vec![0u8; output_len];
    hasher.finalize_xof().read(&mut output);
    output
}

/// SHA3-KMAC128 for passcode (internal use)
pub fn sha3_kmac128_for_passcode(key: &[u8], data: &[u8]) -> Vec<u8> {
    kmac128(key, b"authorization", data, 32)
}

/// SHA3-KMAC128 with customizable parameters
pub fn sha3_kmac128(
    key: &[u8],
    customization: &[u8],
    data: &[u8],
    output_len: usize,
) -> Vec<u8> {
    kmac128(key, customization, data, output_len)
}

/// SHA3-KMAC256 for passcode (internal use)
pub fn sha3_kmac256_for_passcode(key: &[u8], data: &[u8]) -> Vec<u8> {
    kmac256(key, b"authorization", data, 32)
}

/// SHA3-KMAC256 with customizable parameters
pub fn sha3_kmac256(
    key: &[u8],
    customization: &[u8],
    data: &[u8],
    output_len: usize,
) -> Vec<u8> {
    kmac256(key, customization, data, output_len)
}
